pkg/api/auth/token_test.go

```diff
@@ -0,0 +1,167 @@
+package auth
+
+import (
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetImpersonateAuthInfo(t *testing.T) {
+	type input struct {
+		token      string
+		kubeconfig string
+	}
+	type output struct {
+		impersonate       string
+		impersonateGroups []string
+		errorMessage      string
+	}
+
+	var testCases = []struct {
+		name     string
+		given    input
+		expected output
+	}{
+		{
+			name: "valid serviceAccount token",
+			given: input{
+				token: "eyJhbGciOiJSUzI1NiIsImtpZCI6IlZ6WmFWS2N5aUs1WFZBV040V3hSYnlFT3pyOWZIN1hrU1B4NWVSeDVMbDQifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJkZWZhdWx0Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6ImFkbWluLXNhLXRva2VuLW56YzI5Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZXJ2aWNlLWFjY291bnQubmFtZSI6ImFkbWluLXNhIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZXJ2aWNlLWFjY291bnQudWlkIjoiMmZkMTU0ZjctMjlmYy00NTUxLTk2NmYtNDY1NTgyNzEyZDIxIiwic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50OmRlZmF1bHQ6YWRtaW4tc2EifQ.BHdo7zKq-dFtlm0OjwFRyypgRqMWmOkx18kGxbm0qLALnWU4GZF2CGqQxTr9IsfEUR3xak9_NnEnsMIZh9yvqpwQgWZPf0LEtWBl7mw04LiQ0syXzY3sLCT7hnGDJv_sIN_zYH3apCDIENjgXQuXcvr1NUAfP9GtwaZ9JlFj1x63HB24EMSSqPpg4NF55XjyHcm6B2JKqSEfkXk1LzMpn_6XFGVv0ic_tN4c88MhXlcSvJemEsicEwgnouSAxqFXnsjM2napsgUlvctMDcXE5D97OlNhQ_EFLrFpMlr4Yo-lKQ9TRVTZ-Mv4w9zV3CoLlmXjmjq_BRgC5MH-43I5Dw",
+			},
+			expected: output{
+				errorMessage: "",
+				impersonate:  "system:serviceaccount:default:admin-sa",
+			},
+		},
+		{
+			name: "invalid serviceAccount token",
+			given: input{
+				token: "invalid",
+			},
+			expected: output{
+				errorMessage: "token contains an invalid number of segments",
+			},
+		},
+		{
+			name: "valid kubeconfig with client-certificate",
+			given: input{
+				kubeconfig: kubeconfigWithClientCert,
+			},
+			expected: output{
+				errorMessage:      "",
+				impersonate:       "kube-admin",
+				impersonateGroups: []string{"system:masters"},
+			},
+		},
+		{
+			name: "valid kubeconfig with token",
+			given: input{
+				kubeconfig: kubeconfigWithToken,
+			},
+			expected: output{
+				errorMessage: "",
+				impersonate:  "system:serviceaccount:default:admin-sa",
+			},
+		},
+		{
+			name: "invalid kubeconfig",
+			given: input{
+				kubeconfig: inValidkubeconfigWithToken,
+			},
+			expected: output{
+				errorMessage: "token contains an invalid number of segments",
+				impersonate:  "system:serviceaccount:default:admin-sa",
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		authInfo, err := buildAuthInfo(tc.given.token, tc.given.kubeconfig)
+		assert.Nil(t, err, "case %q", tc.name)
+		if err != nil {
+			continue
+		}
```

Comment:
err is asserted to be nil so we don't need this check

Reply:
When the error isn't empty, after assert.Nil it still continue to the next step, no need to run the below test when the first step is failed.